drivers/spi/spi_mchp_mec5_qspi.c

```diff
@@ -0,0 +1,729 @@
+		return -EIO;
+	}
+
+	return 0;
+}
+
+#ifdef CONFIG_SPI_ASYNC
+
+/* Async for non-fast path */
+static void mec5_qspi_async_ctx_next(const struct device *dev)
+{
+	const struct mec5_qspi_config *devcfg = dev->config;
+	struct mec_qspi_regs *regs = devcfg->regs;
+	struct mec5_qspi_data *data = dev->data;
+	struct spi_context *ctx = &data->ctx;
+	size_t xlen = 0;
+	uint32_t qflags = MEC5_QSPI_ULDMA_FLAG_START | MEC5_QSPI_ULDMA_FLAG_IEN;
+
+	spi_context_update_tx(ctx, 1u, data->chunk_size);
+	spi_context_update_rx(ctx, 1u, data->chunk_size);
+
+	if (data->total_tx_size) {
+		data->total_tx_size -= data->chunk_size;
+	}
+
+	if (data->total_rx_size) {
+		data->total_rx_size -= data->chunk_size;
+	}
+
+	if (spi_context_rx_on(ctx) || spi_context_tx_on(ctx)) {
+		xlen = spi_context_max_continuous_chunk(ctx);
+		data->chunk_size = xlen;
+
+		uint8_t const *txb = ctx->tx_buf;
+		uint8_t *rxb = ctx->rx_buf;
+
+		if (txb) {
+			qflags |= MEC5_QSPI_ULDMA_FLAG_INCR_TX;
+		} else {
+			txb = &devcfg->ovrc;
+		}
+
+		if (rxb) {
+			qflags |= MEC5_QSPI_ULDMA_FLAG_INCR_RX;
+		} else {
+			rxb = (uint8_t *)&data->rxdb;
+		}
+
+		if ((data->total_tx_size <= xlen) && (data->total_rx_size <= xlen)) {
+			qflags |= MEC5_QSPI_ULDMA_FLAG_CLOSE;
+		}
+		data->xfr_flags = MEC5_QSPI_XFR_FLAG_LDMA;
+		mec_hal_qspi_uldma_fd2(regs, (const uint8_t *)txb, rxb, xlen, qflags);
+	} else {
+		spi_context_complete(&data->ctx, dev, 0);
+	}
+}
+
+/* Async for fast path */
+static void mec5_qspi_async_next(const struct device *dev)
+{
+	const struct mec5_qspi_config *devcfg = dev->config;
+	struct mec_qspi_regs *regs = devcfg->regs;
+	struct mec5_qspi_data *data = dev->data;
+	uint32_t xflags = MEC5_QSPI_ULDMA_FLAG_START | MEC5_QSPI_ULDMA_FLAG_IEN;
+
+	if (!(data->xfr_flags & MEC5_QSPI_XFR_FLAG_FAST)) {
+		return mec5_qspi_async_ctx_next(dev);
+	}
+
+	if (data->rxcnt) {
+		data->rxcnt--;
+		data->rxb++;
+	}
+	if (data->txcnt) {
+		data->txcnt--;
+		data->txb++;
+	}
+
+	if (!data->rxcnt && !data->txcnt) {
+		spi_context_complete(&data->ctx, dev, 0);
+		return;
+	}
+
+	if ((data->txcnt <= 1) && (data->rxcnt <= 1)) {
+		xflags |= MEC5_QSPI_ULDMA_FLAG_CLOSE;
+	}
+
+	data->xfr_flags = MEC5_QSPI_XFR_FLAG_LDMA | MEC5_QSPI_XFR_FLAG_FAST;
+	mec_hal_qspi_uldma_fd(regs, data->txb->buf, data->txb->len, data->rxb->buf, data->rxb->len,
+			      xflags);
+}
+#endif /* CONFIG_SPI_ASYNC */
+
+static void mec5_qspi_isr(const struct device *dev)
+{
+	struct mec5_qspi_data *data = dev->data;
+	const struct mec5_qspi_config *devcfg = dev->config;
+	struct mec_qspi_regs *regs = devcfg->regs;
+	uint32_t hwsts = 0u;
+	int status = 0;
+
+	hwsts = mec_hal_qspi_hw_status(regs);
+	data->qstatus = hwsts;
+	status = mec_hal_qspi_done(regs);
+
+	mec_hal_qspi_intr_ctrl(regs, 0);
+	mec_hal_qspi_hw_status_clr(regs, hwsts);
+	mec_hal_qspi_girq_clr(regs);
+
+	status = (status == MEC_RET_OK) ? 0 : -EIO;
+
+	if (status == MEC_RET_OK) {
+		status = 0;
+	} else {
+		status = -EIO;
+	}
+
+#ifdef CONFIG_SPI_ASYNC
```

Comment:
Async has nothing to do about running in interrupt mode or not, it's just for the caller: 
- spi_transceive() is blocking until it's finished
- spi_transceive_signal() returns right away, and the result of the transaction will come from a callback.

Reply:
@tbursztyka OK. Will re-architect the driver.

Reply:
Done